test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs

```diff
@@ -0,0 +1,280 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Cardano.Wallet.CoinSelection.LargestFirstSpec
+    ( spec
+    ) where
+
+import Prelude
+
+import Cardano.Wallet.CoinSelection
+    ( CoinSelection (..), CoinSelectionError (..), CoinSelectionOptions (..) )
+import Cardano.Wallet.CoinSelection.LargestFirst
+    ( largestFirst )
+import Cardano.Wallet.Primitive.Types
+    ( Address (..)
+    , Coin (..)
+    , Hash (..)
+    , TxIn (..)
+    , TxOut (..)
+    , UTxO (..)
+    , invariant
+    )
+import Control.DeepSeq
+    ( deepseq )
+import Control.Monad.IO.Class
+    ( liftIO )
+import Control.Monad.Trans.Except
+    ( runExceptT )
+import Data.List.NonEmpty
+    ( NonEmpty (..) )
+import Data.Word
+    ( Word64 )
+import Test.Hspec
+    ( Spec, describe, it, shouldBe )
+import Test.QuickCheck
+    ( Arbitrary (..)
+    , Property
+    , choose
+    , oneof
+    , property
+    , scale
+    , suchThat
+    , vectorOf
+    )
+import Test.QuickCheck.Monadic
+    ( monadicIO )
+
+import qualified Data.List as L
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Map.Strict as Map
+
+
+spec :: Spec
+spec = do
+    describe "Coin selection : LargestFirst algorithm" $ do
+        it "works as expected for zero-fee and fully covered by utxo coin selection"
+            (property propLargestFirstFullyCovered)
+        it "works as expected for zero-fee and not covered by utxo coin selection"
+            (property propLargestFirstNotCovered)
+        it "works as expected for zero-fee and fully covered by utxo coin selection when maximumNumberOfInputs is small"
+            (property propLargestFirstFullyCoveredSmallMaxInput)
```

Comment:
I am a puzzled by the properties here :thinking: ? What properties are we actually looking for? I think you're mis-using property testing here. The testing logic is almost as complex as the implementation itself which is a good sign that tests are maybe doing too much. A good indicator is the way you can express the property in prose, what _means_ "works as expected" here, if you're checking for a property, then you should be able to express it.

I'd suggest the following approach:

- Reword these test cases into plain unit tests, where, given some inputs (that you can construct by hand, in a easy-to-read fashion), you obtain some expected output / behavior.

- Add some property-based tests for actual _properties_ of that largest first heuristic, for instance, I can see a few properties that I'd expect on such function:
    - forall (UTxO, NonEmpty TxOut), running largest first twice one the same UTxO yields exactly the same result
    - forall (UTXO, NonEmpty TxOut), for all selected input, there's no bigger input in the UTxO that is not already in the selected inputs. 
    - forall (UTxO, NonEmpty TxOut), there's at least as many selected inputs as there are requested outputs

So in practice, the properties are very short to express, and it's really about performing the coin selection, and then, verifying that the output satisfies some conditions. All properties above assume that there are enough inputs to cover for all of the NonEmpty TxOut, so likely, they'll need preconditions (which could just be `isRight $ runExceptT $ largestFirst ... ==> ...`)

Reply:
fair argument. Added unit tests already. Properties will be added in the next commit

Reply:
this one is I do not understand  : `forall (UTXO, NonEmpty TxOut), for all selected input, there's no bigger input in the UTxO that is not already in the selected inputs.`

algo goes : we take **maxNumberOfInputs** biggest utxo entries, and then take from the smallest to cover transaction outputs taking from the smallest. It seems, but I may be wrong, that it contradicts ^^^

Reply:
I sticked to `CoveringCase` to run it quicker, as correct subset is significantly smaller to whole set of (Utxo, Empty TxOut)

Reply:
> algo goes : we take maxNumberOfInputs biggest utxo entries, 

Yes.

>  and then take from the smallest to cover transaction outputs taking from the smallest

No. We do cover biggest outputs first, and use the biggest inputs first to do so.

Reply:
right !!!

Reply:
added the property. in the lights of correct implementation it makes sense